### PR TITLE
New Approach for kNN Patching

### PR DESCRIPTION
**Pull request by owerbat**

### CODE REVIEW

File: daal4py/sklearn/neighbors/_base.py [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+    else:
+        predict_alg = d4p.kdtree_knn_classification_prediction
+        compute_fptype = fptype
+
+    params['fptype'] = compute_fptype
+
+    return predict_alg(**params)
+
+
+def parse_auto_method(clf, method, n_samples, n_features):
```
Comment (Alexsandruss) [inline daal4py/sklearn/neighbors/_base.py:65]: `clf` name is inconsistent while `estimator` is used anywhere else

File: daal4py/sklearn/neighbors/_base.py [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+                raise ValueError("Training data only contain information about one class.")
+        else:
+            self._y = y
+
+        try:
+            fptype = getFPType(X)
+        except ValueError:
+            fptype = None
+
+        if daal_check_version((2020, 3)) and not X_incorrect_type \
```
Comment (Alexsandruss) [inline daal4py/sklearn/neighbors/_base.py:244]: Update version check to support oneDAL

File: daal4py/sklearn/neighbors/_base.py [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+
+class KNeighborsMixin(BaseKNeighborsMixin):
+    def kneighbors(self, X=None, n_neighbors=None, return_distance=True):
+        x = self._fit_X if X is None else X
+        try:
+            fptype = getFPType(x)
+        except ValueError:
+            fptype = None
+
+        if daal_check_version((2020, 3)) and hasattr(self, 'daal_model_') \
```
Comment (Alexsandruss) [inline daal4py/sklearn/neighbors/_base.py:266]: Update version check to support oneDAL

File: daal4py/sklearn/neighbors/_classification.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+# daal4py KNN classification scikit-learn-compatible classes
+
+from ._base import NeighborsBase, KNeighborsMixin
+from sklearn.base import ClassifierMixin as BaseClassifierMixin
+from sklearn.neighbors._classification import KNeighborsClassifier as BaseKNeighborsClassifier
+from sklearn.utils.validation import _deprecate_positional_args
+from sklearn.neighbors._base import _check_weights
+
+
+class KNeighborsClassifier(NeighborsBase, KNeighborsMixin, BaseClassifierMixin):
```
Comment (Alexsandruss) [inline daal4py/sklearn/neighbors/_classification.py:27]: Isn't inference from sklearn kNeighborsClassifier better?

File: daal4py/sklearn/neighbors/_base.py [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+    if return_distance:
+        params['resultsToCompute'] += '|computeDistances'
+
+    method = parse_auto_method(estimator, estimator._fit_method, estimator.n_samples_fit_, n_features)
+
+    predict_alg = prediction_algorithm(method, fptype, params)
+    prediction_result = predict_alg.compute(X, estimator.daal_model_)
+
+    if return_distance:
+        results = prediction_result.distances.astype(fptype), prediction_result.indices.astype(int)
```
Comment (Alexsandruss) [inline daal4py/sklearn/neighbors/_base.py:164]: For some reason, kd_tree always returns distances as float64, brute - depends on X from fit().

File: daal4py/sklearn/neighbors/_base.py [hunk 6]
```diff
@@ -0,0 +1,416 @@
...
+                if SKLEARN_24:
+                    BaseNeighborsBase._fit(self, self._fit_X, self._y)
+                else:
+                    BaseNeighborsBase._fit(self, self._fit_X)
+            result = super(KNeighborsMixin, self).kneighbors(X, n_neighbors, return_distance)
+
+        return result
+
+
+class RadiusNeighborsMixin(BaseRadiusNeighborsMixin):
```
Comment (PetrovKP) [inline daal4py/sklearn/neighbors/_base.py:373]: it's necessary?

File: daal4py/sklearn/neighbors/_base.py [hunk 7]
```diff
@@ -0,0 +1,416 @@
...
+    if n_features and shape and len(shape) > 1 and shape[1] != n_features:
+        raise ValueError('Input data shape {} is inconsistent with the trained model'.format(X.shape))
+
+    try:
+        fptype = getFPType(X)
+    except ValueError:
+        fptype = None
+
+    if daal_check_version(((2020,'P', 3),(2021,'B', 110))) and not (daal_model is None) \
+    and fptype is not None and not sp.issparse(X):
```
Comment (PetrovKP) [inline daal4py/sklearn/neighbors/_base.py:396]: I think should be enough daal_model is None

File: daal4py/sklearn/neighbors/_base.py [hunk 8]
```diff
@@ -0,0 +1,416 @@
...
+
+class RadiusNeighborsMixin(BaseRadiusNeighborsMixin):
+    def radius_neighbors(self, X=None, radius=None, return_distance=True,
+                         sort_results=False):
+        if getattr(self, '_tree', 0) is None and self._fit_method == 'kd_tree':
+            raise ValueError('oneDAL does not build sklearn.neighbors._kd_tree.KDTree')
+        return BaseRadiusNeighborsMixin.radius_neighbors(self, X, radius, return_distance, sort_results)
+
+
+def daal4py_classifier_predict(estimator, X, base_predict):
```
Comment (PetrovKP) [inline daal4py/sklearn/neighbors/_base.py:381]: It should be transferred to the next file.

**Review by PetrovKP — COMMENTED** [review]
I like in generally

File: daal4py/sklearn/neighbors/_base.py (hunk 7 above)
Comment (owerbat) [reply daal4py/sklearn/neighbors/_base.py]: Fixed

File: daal4py/sklearn/neighbors/_base.py [hunk 9]
```diff
@@ -0,0 +1,416 @@
...
+    def kneighbors(self, X=None, n_neighbors=None, return_distance=True):
+        daal_model = getattr(self, '_daal_model', None)
+        x = self._fit_X if X is None else X
+        try:
+            fptype = getFPType(x)
+        except ValueError:
+            fptype = None
+
+        if daal_check_version(((2020,'P', 3),(2021,'B', 110))) and not (daal_model is None) \
+        and fptype is not None and not sp.issparse(X):
```
[the remark this answers is not in the document]
Comment (owerbat) [reply daal4py/sklearn/neighbors/_base.py:354]: Fixed

**Review by PetrovKP — APPROVED** [review]

**Pull request merged**